File: sdks/javascript/src/mount.ts
/**
 * ObjectFS Mount Manager
 *
 * Handles mounting and unmounting of ObjectFS filesystems.
 */

import { spawn, ChildProcess } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { promisify } from 'util';
import { Configuration } from './config';
import { MountError, ConfigurationError } from './errors';
import { MountOptions, UnmountOptions, MountInfo } from './types';

const writeFile = promisify(fs.writeFile);
const unlink = promisify(fs.unlink);
const access = promisify(fs.access);
const mkdir = promisify(fs.mkdir);
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

export class MountManager {
  constructor(
    private binaryPath: string,
    private config: Configuration
  ) {}

  /**
   * Mount ObjectFS filesystem
   */
  async mount(
    storageUri: string,
    mountPoint: string,
    config: Configuration,
    options: MountOptions = {}
  ): Promise<ChildProcess> {
    // Validate inputs
    this.validateMountInputs(storageUri, mountPoint);

    // Prepare mount point
    await this.prepareMountPoint(mountPoint);

    // Generate configuration file if needed
    let configFile: string | undefined;
    try {
      configFile = await this.createTempConfig(config);

      // Build command
      const cmd = this.buildMountCommand(
        storageUri,
        mountPoint,
        configFile,
        options.foreground || false
      );

      console.log(`Mounting ${storageUri} at ${mountPoint}`);
      console.log(`Mount command: ${cmd.join(' ')}`);

      // Start mount process
      const process = spawn(cmd[0], cmd.slice(1), {
        stdio: options.foreground ? 'inherit' : 'pipe',
      });

      if (options.foreground) {
        // For foreground mounts, wait for completion
        return new Promise((resolve, reject) => {
          process.on('close', code => {
            if (code === 0) {
              resolve(process);
            } else {
              reject(new MountError(`Mount failed with code ${code}`));
            }
          });

          process.on('error', error => {
            reject(new MountError(`Mount process error: ${error}`));
          });
        });
      } else {
        // For background mounts, wait for mount to be ready
        await this.waitForMount(mountPoint, options.timeout || 30000);
        return process;
      }
    } catch (error) {
      if (configFile) {
        try {
          await unlink(configFile);
        } catch (unlinkError) {
          // Ignore cleanup errors
        }
      }
      throw new MountError(`Failed to mount ${storageUri}: ${error}`);
    }
  }

  /**
   * Unmount ObjectFS filesystem
   */
  async unmount(
    mountPoint: string,
    options: UnmountOptions = {}
  ): Promise<boolean> {
    if (!(await this.isMounted(mountPoint))) {
      console.warn(`${mountPoint} is not mounted`);
      return true;
    }

    try {
      const timeout = options.timeout || 10000;

      // Try graceful unmount first
      const cmd = ['fusermount', '-u'];
      if (options.force) {
        cmd.push('-z'); // Lazy unmount
      }
      cmd.push(mountPoint);

      console.log(`Unmounting ${mountPoint}`);

      const process = spawn(cmd[0], cmd.slice(1), {
        stdio: 'pipe',
      });

      const result = await this.waitForProcess(process, timeout);

      if (result.code === 0) {
        // Wait for unmount to complete
        await this.waitForUnmount(mountPoint, timeout);
        console.log(`Successfully unmounted ${mountPoint}`);
        return true;
      } else {
        console.error(`Unmount failed: ${result.stderr}`);
        if (options.force) {
          return this.forceUnmount(mountPoint);
        }
        return false;
      }
    } catch (error) {
      console.error(`Unmount error for ${mountPoint}: ${error}`);
      return false;
    }
  }

  /**
   * Check if directory is mounted with ObjectFS
   */
  async isMounted(mountPoint: string): Promise<boolean> {
    try {
      const absolutePath = path.resolve(mountPoint);

      // Check if mount point exists and is a directory
      try {
        const stats = await stat(absolutePath);
        if (!stats.isDirectory()) {
          return false;
        }
      } catch (error) {
        return false;
      }

      // Check system mount table (Linux-specific)
      if (os.platform() === 'linux') {
        try {
          const mounts = fs.readFileSync('/proc/mounts', 'utf8');
          const lines = mounts.split('\n');

          for (const line of lines) {
            const parts = line.split(' ');
            if (parts.length >= 3 && parts[1] === absolutePath) {
              // Check if it's a FUSE mount (ObjectFS uses FUSE)
              return parts[2] === 'fuse' || parts[0].includes('objectfs');
            }
          }
        } catch (error) {
          // Fall back to other detection methods
        }
      }

      // Check using mount command (cross-platform)
      try {
        const { spawn } = require('child_process');
        const process = spawn('mount', [], { stdio: 'pipe' });

        return new Promise<boolean>((resolve) => {
          let output = '';
          process.stdout.on('data', (data: Buffer) => {
            output += data.toString();
          });

          process.on('close', () => {
            const lines = output.split('\n');
            for (const line of lines) {
              if (line.includes(absolutePath) &&
                  (line.includes('fuse') || line.includes('objectfs'))) {
                resolve(true);
                return;
              }
            }
            resolve(false);
          });

          process.on('error', () => resolve(false));
        });
      } catch (error) {
        console.debug(`Error checking mount status for ${mountPoint}: ${error}`);
        return false;
      }

      return false;
    } catch (error) {
      console.debug(`Error checking mount status for ${mountPoint}: ${error}`);
      return false;
    }
  }

  /**
   * List all ObjectFS mounts
   */
  async listMounts(): Promise<MountInfo[]> {
    const mounts: MountInfo[] = [];

    try {
      // Read mount information from /proc/mounts on Linux
      if (os.platform() === 'linux') {
        try {
          const mountsContent = fs.readFileSync('/proc/mounts', 'utf8');
          const lines = mountsContent.split('\n');

          for (const line of lines) {
            const parts = line.split(' ');
            if (parts.length >= 4 &&
                (parts[2] === 'fuse' || parts[0].includes('objectfs'))) {

              const mountInfo: MountInfo = {
                device: parts[0],
                mountpoint: parts[1],
                fstype: parts[2],
                opts: parts[3],
              };

              // Add usage statistics if available
              try {
                const stats = fs.statSync(parts[1]);
                if (stats.isDirectory()) {
                  // Try to get disk usage (simplified)
                  mountInfo.total = 0;
                  mountInfo.used = 0;
                  mountInfo.free = 0;
                  mountInfo.percent = 0;
                }
              } catch (error) {
                // Ignore errors getting usage stats
              }

              mounts.push(mountInfo);
            }
          }
        } catch (error) {
          console.error('Error reading /proc/mounts:', error);
        }
      }

      // Fallback: use mount command
      if (mounts.length === 0) {
        try {
          const { spawn } = require('child_process');
          const process = spawn('mount', [], { stdio: 'pipe' });

          await new Promise<void>((resolve) => {
            let output = '';
            process.stdout.on('data', (data: Buffer) => {
              output += data.toString();
            });

            process.on('close', () => {
              const lines = output.split('\n');
              for (const line of lines) {
                if (line.includes('fuse') || line.includes('objectfs')) {
                  // Parse mount line: device on mountpoint type fstype (opts)
                  const match = line.match(/^(.+?) on (.+?) type (.+?) \((.+?)\)$/);
                  if (match) {
                    mounts.push({
                      device: match[1],
                      mountpoint: match[2],
                      fstype: match[3],
                      opts: match[4],
                    });
                  }
                }
              }
              resolve();
            });

            process.on('error', () => resolve());
          });
        } catch (error) {
          console.error('Error listing mounts:', error);
        }
      }
    } catch (error) {
      console.error('Error listing mounts:', error);
    }

    return mounts;
  }

  /**
   * Get detailed information about a specific mount
   */
  async getMountInfo(mountPoint: string): Promise<MountInfo | null> {
    const absolutePath = path.resolve(mountPoint);
    const mounts = await this.listMounts();

    return mounts.find(mount => mount.mountpoint === absolutePath) || null;
  }

  // Private helper methods

  private validateMountInputs(storageUri: string, mountPoint: string): void {
    if (!storageUri) {
      throw new MountError('Storage URI cannot be empty');
    }

    // Validate storage URI format
    if (!storageUri.match(/^(s3|gs|az):\/\/.+/)) {
      throw new MountError(`Unsupported storage URI: ${storageUri}`);
    }

    // Validate mount point
    const parentDir = path.dirname(path.resolve(mountPoint));
    if (!fs.existsSync(parentDir)) {
      throw new MountError(`Mount point parent directory does not exist: ${parentDir}`);
    }
  }

  private async prepareMountPoint(mountPoint: string): Promise<void> {
    try {
      const absolutePath = path.resolve(mountPoint);

      // Create directory if it doesn't exist
      await mkdir(absolutePath, { recursive: true });

      // Check if directory is empty
      try {
        const files = await readdir(absolutePath);
        if (files.length > 0) {
          console.warn(`Mount point ${absolutePath} is not empty`);
        }
      } catch (error) {
        // Ignore readdir errors
      }

      // Check permissions
      try {
        await access(absolutePath, fs.constants.R_OK | fs.constants.W_OK);
      } catch (error) {
        throw new MountError(`Insufficient permissions for mount point: ${absolutePath}`);
      }
    } catch (error) {
      if (error instanceof MountError) throw error;
      throw new MountError(`Failed to prepare mount point ${mountPoint}: ${error}`);
    }
  }

  private async createTempConfig(config: Configuration): Promise<string> {
    try {
      config.validate();

      const tempDir = os.tmpdir();
      const configPath = path.join(tempDir, `objectfs-${Date.now()}.yaml`);

      await writeFile(configPath, config.toYAML(), 'utf8');
      return configPath;
    } catch (error) {
      throw new ConfigurationError(`Failed to create configuration file: ${error}`);
    }
  }

  private buildMountCommand(
    storageUri: string,
    mountPoint: string,
    configFile: string,
    foreground: boolean
  ): string[] {
    const cmd = [this.binaryPath];

    // Add configuration file
    if (configFile) {
      cmd.push('--config', configFile);
    }

    // Add foreground flag
    if (foreground) {
      cmd.push('--foreground');
    }

    // Add log level
    cmd.push('--log-level', this.config.global.logLevel);

    // Add storage URI and mount point
    cmd.push(storageUri, mountPoint);

    return cmd;
  }

  private async waitForMount(mountPoint: string, timeout: number): Promise<void> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      if (await this.isMounted(mountPoint)) {
        // Additional check: try to access the mount point
        try {
          await readdir(mountPoint);
          return;
        } catch (error) {
          // Continue waiting
        }
      }

      await new Promise(resolve => setTimeout(resolve, 100));
    }

    throw new MountError(`Mount timeout after ${timeout}ms`);
  }

  private async waitForUnmount(mountPoint: string, timeout: number): Promise<void> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      if (!(await this.isMounted(mountPoint))) {
        return;
      }
      await new Promise(resolve => setTimeout(resolve, 100));
    }

    throw new MountError(`Unmount timeout after ${timeout}ms`);
  }

  private async forceUnmount(mountPoint: string): Promise<boolean> {
    try {
      // Try lazy unmount
      const process = spawn('fusermount', ['-u', '-z', mountPoint], {
        stdio: 'pipe',
      });

      const result = await this.waitForProcess(process, 10000);

      if (result.code === 0) {
        console.log(`Force unmount successful for ${mountPoint}`);
        return true;
      } else {
        console.error(`Force unmount failed: ${result.stderr}`);
        return false;
      }
    } catch (error) {
      console.error(`Force unmount error: ${error}`);
      return false;
    }
  }

  private async waitForProcess(
    process: ChildProcess,
    timeout: number
  ): Promise<{ code: number | null; stdout: string; stderr: string }> {
    return new Promise((resolve, reject) => {
      let stdout = '';
      let stderr = '';

      const timeoutId = setTimeout(() => {
        process.kill('SIGKILL');
        reject(new Error(`Process timeout after ${timeout}ms`));
      }, timeout);

      if (process.stdout) {
        process.stdout.on('data', (data: Buffer) => {
          stdout += data.toString();
        });
      }

      if (process.stderr) {
        process.stderr.on('data', (data: Buffer) => {
          stderr += data.toString();
        });
      }

      process.on('close', (code) => {
        clearTimeout(timeoutId);
        resolve({ code, stdout, stderr });
      });

      process.on('error', (error) => {
        clearTimeout(timeoutId);
        reject(error);
      });
    });
  }
}

File: sdks/python/objectfs/mount.py
"""
ObjectFS Mount Manager

Handles mounting and unmounting of ObjectFS filesystems.
"""

import logging
import os
import psutil
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Dict, List, Optional, Union, Any

from .config import Configuration
from .exceptions import MountError, ConfigurationError

logger = logging.getLogger(__name__)


class MountManager:
    """
    Manages ObjectFS filesystem mounts.

    Provides functionality to mount, unmount, and monitor ObjectFS
    filesystem instances with comprehensive error handling and validation.
    """

    def __init__(self, binary_path: str, config: Configuration):
        """
        Initialize mount manager.

        Args:
            binary_path: Path to ObjectFS binary
            config: ObjectFS configuration
        """
        self.binary_path = binary_path
        self.config = config

    def mount(
        self,
        storage_uri: str,
        mount_point: Union[str, Path],
        config: Optional[Configuration] = None,
        foreground: bool = False,
        timeout: int = 30
    ) -> subprocess.Popen:
        """
        Mount ObjectFS filesystem.

        Args:
            storage_uri: Storage URI (e.g., s3://bucket-name)
            mount_point: Local mount point directory
            config: Configuration overrides
            foreground: Run in foreground mode
            timeout: Mount timeout in seconds

        Returns:
            Process object for the mount

        Raises:
            MountError: If mount operation fails
        """
        mount_point = Path(mount_point).resolve()
        effective_config = config or self.config

        # Validate inputs
        self._validate_mount_inputs(storage_uri, mount_point)

        # Prepare mount point
        self._prepare_mount_point(mount_point)

        # Generate configuration file if needed
        config_file = None
        try:
            config_file = self._create_temp_config(effective_config)

            # Build command
            cmd = self._build_mount_command(
                storage_uri, mount_point, config_file, foreground
            )

            logger.info(f"Mounting {storage_uri} at {mount_point}")
            logger.debug(f"Mount command: {' '.join(cmd)}")

            # Start mount process
            process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                universal_newlines=True
            )

            if foreground:
                # For foreground mounts, wait for completion
                stdout, stderr = process.communicate()
                if process.returncode != 0:
                    raise MountError(f"Mount failed: {stderr}")
                return process
            else:
                # For background mounts, wait for mount to be ready
                self._wait_for_mount(mount_point, timeout)
                return process

        except Exception as e:
            if config_file and os.path.exists(config_file):
                os.unlink(config_file)
            raise MountError(f"Failed to mount {storage_uri}: {e}") from e

    def unmount(
        self,
        mount_point: Union[str, Path],
        force: bool = False,
        timeout: int = 10
    ) -> bool:
        """
        Unmount ObjectFS filesystem.

        Args:
            mount_point: Mount point to unmount
            force: Force unmount if busy
            timeout: Unmount timeout in seconds

        Returns:
            True if successfully unmounted

        Raises:
            MountError: If unmount operation fails
        """
        mount_point = Path(mount_point).resolve()

        if not self.is_mounted(mount_point):
            logger.warning(f"{mount_point} is not mounted")
            return True

        try:
            # Try graceful unmount first
            cmd = ['fusermount', '-u', str(mount_point)]
            if force:
                cmd.insert(1, '-z')  # Lazy unmount

            logger.info(f"Unmounting {mount_point}")
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout
            )

            if result.returncode == 0:
                # Wait for unmount to complete
                self._wait_for_unmount(mount_point, timeout)
                logger.info(f"Successfully unmounted {mount_point}")
                return True
            else:
                logger.error(f"Unmount failed: {result.stderr}")
                if force:
                    return self._force_unmount(mount_point)
                return False

        except subprocess.TimeoutExpired:
            logger.error(f"Unmount timed out for {mount_point}")
            if force:
                return self._force_unmount(mount_point)
            return False
        except Exception as e:
            logger.error(f"Unmount error for {mount_point}: {e}")
            return False

    def is_mounted(self, mount_point: Union[str, Path]) -> bool:
        """
        Check if directory is mounted with ObjectFS.

        Args:
            mount_point: Directory to check

        Returns:
            True if mounted with ObjectFS
        """
        mount_point = Path(mount_point).resolve()

        try:
            # Check if mount point exists and is a directory
            if not mount_point.exists() or not mount_point.is_dir():
                return False

            # Check system mount table
            for mount in psutil.disk_partitions():
                if mount.mountpoint == str(mount_point):
                    # Check if it's a FUSE mount (ObjectFS uses FUSE)
                    return mount.fstype == 'fuse' or 'objectfs' in mount.device

            return False

        except Exception as e:
            logger.debug(f"Error checking mount status for {mount_point}: {e}")
            return False

    def list_mounts(self) -> List[Dict[str, Any]]:
        """
        List all ObjectFS mounts.

        Returns:
            List of mount information dictionaries
        """
        mounts = []

        try:
            for mount in psutil.disk_partitions():
                if mount.fstype == 'fuse' or 'objectfs' in mount.device:
                    mount_info = {
                        'device': mount.device,
                        'mountpoint': mount.mountpoint,
                        'fstype': mount.fstype,
                        'opts': mount.opts
                    }

                    # Add usage statistics if available
                    try:
                        usage = psutil.disk_usage(mount.mountpoint)
                        mount_info.update({
                            'total': usage.total,
                            'used': usage.used,
                            'free': usage.free,
                            'percent': usage.percent
                        })
                    except Exception:
                        pass

                    mounts.append(mount_info)

        except Exception as e:
            logger.error(f"Error listing mounts: {e}")

        return mounts

    def get_mount_info(self, mount_point: Union[str, Path]) -> Optional[Dict[str, Any]]:
        """
        Get detailed information about a specific mount.

        Args:
            mount_point: Mount point to query

        Returns:
            Mount information dictionary or None if not found
        """
        mount_point = Path(mount_point).resolve()

        for mount in self.list_mounts():
            if mount['mountpoint'] == str(mount_point):
                return mount

        return None

    # Private helper methods

    def _validate_mount_inputs(self, storage_uri: str, mount_point: Path):
        """Validate mount inputs."""
        if not storage_uri:
            raise MountError("Storage URI cannot be empty")

        # Validate storage URI format
        if not storage_uri.startswith(('s3://', 'gs://', 'az://')):
            raise MountError(f"Unsupported storage URI: {storage_uri}")

        # Validate mount point
        if not mount_point.parent.exists():
            raise MountError(f"Mount point parent directory does not exist: {mount_point.parent}")

    def _prepare_mount_point(self, mount_point: Path):
        """Prepare mount point directory."""
        try:
            mount_point.mkdir(parents=True, exist_ok=True)

            # Check if directory is empty
            if any(mount_point.iterdir()):
                logger.warning(f"Mount point {mount_point} is not empty")

            # Check permissions
            if not os.access(mount_point, os.R_OK | os.W_OK):
                raise MountError(f"Insufficient permissions for mount point: {mount_point}")

        except PermissionError:
            raise MountError(f"Permission denied creating mount point: {mount_point}")
        except Exception as e:
            raise MountError(f"Failed to prepare mount point {mount_point}: {e}")

    def _create_temp_config(self, config: Configuration) -> str:
        """Create temporary configuration file."""
        try:
            config.validate()

            fd, config_path = tempfile.mkstemp(suffix='.yaml', prefix='objectfs-')
            with os.fdopen(fd, 'w') as f:
                f.write(config.to_yaml())

            return config_path

        except Exception as e:
            raise ConfigurationError(f"Failed to create configuration file: {e}")

    def _build_mount_command(
        self,
        storage_uri: str,
        mount_point: Path,
        config_file: str,
        foreground: bool
    ) -> List[str]:
        """Build mount command arguments."""
        cmd = [self.binary_path]

        # Add configuration file
        if config_file:
            cmd.extend(['--config', config_file])

        # Add foreground flag
        if foreground:
            cmd.append('--foreground')

        # Add log level
        cmd.extend(['--log-level', self.config.global_config.log_level])

        # Add storage URI and mount point
        cmd.extend([storage_uri, str(mount_point)])

        return cmd

    def _wait_for_mount(self, mount_point: Path, timeout: int):
        """Wait for mount to be ready."""
        start_time = time.time()

        while time.time() - start_time < timeout:
            if self.is_mounted(mount_point):
                # Additional check: try to access the mount point
                try:
                    list(mount_point.iterdir())
                    return
                except Exception:
                    pass

            time.sleep(0.1)

        raise MountError(f"Mount timeout after {timeout} seconds")

    def _wait_for_unmount(self, mount_point: Path, timeout: int):
        """Wait for unmount to complete."""
        start_time = time.time()

        while time.time() - start_time < timeout:
            if not self.is_mounted(mount_point):
                return
            time.sleep(0.1)

        raise MountError(f"Unmount timeout after {timeout} seconds")

    def _force_unmount(self, mount_point: Path) -> bool:
        """Force unmount using system commands."""
        try:
            # Try lazy unmount
            result = subprocess.run(
                ['fusermount', '-u', '-z', str(mount_point)],
                capture_output=True,
                text=True,
                timeout=10
            )

            if result.returncode == 0:
                logger.info(f"Force unmount successful for {mount_point}")
                return True
            else:
                logger.error(f"Force unmount failed: {result.stderr}")
                return False

        except Exception as e:
            logger.error(f"Force unmount error: {e}")
            return False
